src/client/mapblock_mesh.cpp

```diff
@@ -1003,6 +1004,164 @@ static void applyTileColor(PreMeshBuffer &pmb)
 	}
 }
 
+/*
+	MapBlockBspTree
+*/
+
+void MapBlockBspTree::buildTree(const std::vector<MeshTriangle> *triangles)
+{
+	this->triangles = triangles;
+
+	nodes.clear();
+
+	// assert that triangle index can fit into s32
+	assert(triangles->size() <= 0x7FFFFFFFL);
+	std::vector<s32> indexes;
+	indexes.reserve(triangles->size());
+	for (u32 i = 0; i < triangles->size(); i++)
+		indexes.push_back(i);
+
+	root = buildTree(v3f(1, 0, 0), v3f(85, 85, 85), 40, indexes, 0);
+}
+
+static const MeshTriangle *findSplitCandidate(const std::vector<s32> &list, const std::vector<MeshTriangle> &triangles)
+{
+	// find best origin as a center of the cluster.
+	v3f origin(0, 0, 0);
+	size_t n = list.size();
+	for (s32 i : list) {
+		origin += triangles[i].centroid / n;
+	}
+
+	// find the triangle with the largest area and closest to the center
+	const MeshTriangle *candidate_triangle = &triangles[list[0]];
+	const MeshTriangle *ith_triangle;
+	for (s32 i : list) {
+		ith_triangle = &triangles[i];
```

Comment:
```C++
for (const MeshTriangle &t : list) {
	if (t.areaSQ > candidate_triangle->areaSQ ||
```
The range-based `for` loop is shorter, supersedes an additional variable and `t` still allows address-of operators for the return value.

Reply:
Well, `list` contains indexes of the triangles, and I get the actual triangle from `triangles`. I don't think I an do this in a different way.

Reply:
Oh yes, you are totally right. :+1: